src/cargo/ops/cargo_rustc/engine.rs

```diff
@@ -0,0 +1,138 @@
+use std::collections::HashMap;
+use std::c_str::CString;
+use std::io::process::ProcessOutput;
+use std::fmt::{mod, Show, Formatter};
+
+use util::{mod, CargoResult, ProcessError, ProcessBuilder};
+
+/// Trait for objects that can execute commands.
+pub trait ExecEngine {
```

Comment:
It may be worth having this trait inherit from `Send + Sync` to cut down on writing `Send + Sync` all over the place.